### PR TITLE
Extract ACL checks and add security check endpoint.

### PR DESCRIPTION
**Pull request by Teyras**
Please review this carefully. Some methods could not have been split:

- create assignment
- upload supplementary files
- create/fork exercise
- add group
- add admin (add supervisor already has the ID in path)
- create pipeline
- fork pipeline
- sis create group
- bind group
- submit

### CODE REVIEW

File: app/V1Module/presenters/AssignmentSolutionsPresenter.php [hunk 1]
```diff
@@ -240,27 +263,31 @@ public function actionDownloadSolutionArchive(string $id) {
...
 
+  public function checkDownloadResultArchive(string $id) {
+    $submission = $this->assignmentSolutionSubmissions->findOrThrow($id);
+    if (!$this->assignmentSolutionAcl->canDownloadResultArchive($submission->getAssignmentSolution())) {
+      throw new ForbiddenRequestException("You cannot access result archive for this submission");
+    }
+
+    $this->evaluationLoadingHelper->loadEvaluation($submission);
+
+    if (!$submission->hasEvaluation()) {
```
Comment (Neloop) [inline app/V1Module/presenters/AssignmentSolutionsPresenter.php]: This in my opinion is not authorization check and should be in action method. User which have permissions to see button for downloading archive should see it even if it is not ready yet...

File: app/V1Module/presenters/GroupsPresenter.php [hunk 2]
```diff
@@ -436,6 +468,18 @@ public function actionExercises(string $id) {
     $this->sendSuccessResponse($exercises);
   }
 
+  public function checkStats(string $id) {
+    $group = $this->groups->findOrThrow($id);
+
+    if (!$this->groupAcl->canViewStats($group)) {
+      $user = $this->getCurrentUser();
+
+      if (!($this->groupAcl->canViewStudentStats($group, $user) && $group->isStudentOf($user))) {
```
Comment (Neloop) [inline app/V1Module/presenters/GroupsPresenter.php]: Shouldn't this be negation of condition below on line 451?

File: app/V1Module/presenters/GroupsPresenter.php [hunk 3]
```diff
@@ -512,6 +548,17 @@ public function actionAddStudent(string $id, string $userId) {
     if ($group->isOrganizational()) {
       throw new InvalidArgumentException("It is forbidden to add students to organizational groups");
```
Comment (Neloop) [inline app/V1Module/presenters/GroupsPresenter.php]: Again... not really sure if this should be in authorization check

File: app/V1Module/presenters/SecurityPresenter.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    try {
+      call_user_func_array([$presenter, $methodName], $arguments);
+    } catch (Exception $e) {
+      $result = false;
+    }
+
+    $this->sendSuccessResponse([
+      "result" => $result,
+      "certain" => true
```
Comment (Neloop) [inline app/V1Module/presenters/SecurityPresenter.php]: What exactly certain means?

File: app/V1Module/presenters/GroupsPresenter.php (hunk 2 above)
Comment (Teyras) [reply app/V1Module/presenters/GroupsPresenter.php]: The method should check that the user either a) can view all statistics or b) is a student of the group and can view their own statistics. Is there any discrepancy between this and the code?

File: app/V1Module/presenters/GroupsPresenter.php (hunk 3 above)
Comment (Teyras) [reply app/V1Module/presenters/GroupsPresenter.php]: I'm not certain either. However, I think that if the API call is likely to end with an InvalidArgumentException, we should not give the user false hopes in the check.

File: app/V1Module/presenters/AssignmentSolutionsPresenter.php (hunk 1 above)
Comment (Teyras) [reply app/V1Module/presenters/AssignmentSolutionsPresenter.php]: Yeah, I'll probably give you this one.

File: app/V1Module/presenters/GroupsPresenter.php (hunk 3 above)
Comment (Neloop) [reply app/V1Module/presenters/GroupsPresenter.php]: Well not convinced in either ways... independent opinion by @krulis-martin ?

File: app/V1Module/presenters/GroupsPresenter.php (hunk 2 above)
Comment (Neloop) [reply app/V1Module/presenters/GroupsPresenter.php]: Sorrry... my bad... I did not see parenthesis

File: app/V1Module/presenters/GroupsPresenter.php (hunk 3 above)
Comment (krulis-martin) [reply app/V1Module/presenters/GroupsPresenter.php]: In my opinion, the security endpoint check should follow the behavior of regular endpoint as closely as possible. I.e., if it is forbidden to add students to organizational group, then security should reflect this in authorization.
(Despite the name "security endpoint", I think the actual function is kind of preflight validation.)

Comment (Neloop) [reply app/V1Module/presenters/GroupsPresenter.php]: Ok, that is the thing, I am thinking of it as really just a security endpoint. But if it seems ok to you...

**Pull request merged**